orca/producer_test.go

```diff
@@ -0,0 +1,538 @@
+		panic(fmt.Sprintf("len(addrs) != 1; was: %v", addrs))
+	}
+	sc, err := w.ClientConn.NewSubConn(addrs, opts)
+	if err != nil {
+		return sc, err
+	}
+	l := getListenerAndOptions(addrs[0])
+	l.listener.cleanup = orca.RegisterOOBListener(sc, l.listener, l.opts)
+	l.sc = sc
+	return sc, nil
+}
+
+// listenerAndOptions is stored in an address's attributes to allow ORCA
+// listeners to be registered on subconns created for that address.
+type listenerAndOptions struct {
+	listener *testOOBListener
+	opts     orca.OOBListenerOptions
+	sc       balancer.SubConn // Set by the LB policy
+}
+
+type listenerAndOptionsKey struct{}
+
+func setListenerAndOptions(addr resolver.Address, l *listenerAndOptions) resolver.Address {
+	addr.Attributes = addr.Attributes.WithValue(listenerAndOptionsKey{}, l)
+	return addr
+}
+
+func getListenerAndOptions(addr resolver.Address) *listenerAndOptions {
+	return addr.Attributes.Value(listenerAndOptionsKey{}).(*listenerAndOptions)
+}
+
+// testOOBListener is a simple listener that pushes reports to a channel.
+type testOOBListener struct {
+	cleanup      func()
+	loadReportCh chan *v3orcapb.OrcaLoadReport
+}
+
+func newTestOOBListener() *testOOBListener {
+	return &testOOBListener{cleanup: func() {}, loadReportCh: make(chan *v3orcapb.OrcaLoadReport)}
+}
+
+func (t *testOOBListener) Stop() { t.cleanup() }
+
+func (t *testOOBListener) OnLoadReport(r *v3orcapb.OrcaLoadReport) {
+	t.loadReportCh <- r
+}
+
+// TestProducer is a basic, end-to-end style test of an LB policy with an
+// OOBListener communicating with a server with an ORCA service.
+func (s) TestProducer(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Use a fixed backoff for stream recreation.
+	oldBackoff := internal.DefaultBackoffFunc
+	internal.DefaultBackoffFunc = func(int) time.Duration { return 10 * time.Millisecond }
+	defer func() { internal.DefaultBackoffFunc = oldBackoff }()
+
+	// Initialize listener for our ORCA server.
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Register the OpenRCAService with a very short metrics reporting interval.
+	const shortReportingInterval = 50 * time.Millisecond
+	opts := orca.ServiceOptions{MinReportingInterval: shortReportingInterval}
+	internal.AllowAnyMinReportingInterval.(func(*orca.ServiceOptions))(&opts)
+	s := grpc.NewServer()
+	orcaSrv, err := orca.Register(s, opts)
+	if err != nil {
+		t.Fatalf("orca.Register failed: %v", err)
+	}
+	go s.Serve(lis)
+	defer s.Stop()
+
+	// Create our client with an OOB listener in the LB policy it selects.
+	r := manual.NewBuilderWithScheme("whatever")
+	oobLis := newTestOOBListener()
+
+	lisOpts := orca.OOBListenerOptions{ReportInterval: 50 * time.Millisecond}
+	lao := &listenerAndOptions{listener: oobLis, opts: lisOpts}
+	r.InitialState(resolver.State{Addresses: []resolver.Address{setListenerAndOptions(resolver.Address{Addr: lis.Addr().String()}, lao)}})
```

Comment:
Can you split this line up to maybe read setListenerAndOptions to a local var? This line is long and is hard to read.

Reply:
I don't see this split up still.

Reply:
I created `addr`?